app/download.go

```diff
@@ -34,10 +37,25 @@ func (a *App) DownloadFromURL(downloadURL string) ([]byte, error) {
 	client := a.HTTPService().MakeClient(true)
 	client.Timeout = HTTP_REQUEST_TIMEOUT
 
-	resp, err := client.Get(downloadURL)
+	var resp *http.Response
+	err = utils.ProgressiveRetry(func() error {
+		resp, err = client.Get(downloadURL)
+		respStatusOK := resp.StatusCode >= 200 && resp.StatusCode < 300
+
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch from %s", downloadURL)
+		}
+
+		if !respStatusOK {
```

Comment:
There's a small bug here, because if a 500 response is received, we will continue to retry without closing the body. I'd suggest to rewrite it like:

```go
var resp *http.Response
err = utils.ProgressiveRetry(func() error {
	resp, err = client.Get(downloadURL)
	if err != nil {
		return errors.Wrapf(err, "failed to fetch from %s", downloadURL)
	}

	if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
		_ , _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
		return errors.Errorf("failed to fetch from %s", downloadURL)
	}

	return nil
})
```

Reply:
I want to ask something here. If we are not doing anything with the response body, why do we need to copy it? Copying to `io.Discard` basically does nothing. Why couldn't we do `resp.Body.Close()` directly?

Reply:
Sure, great question. It is a very common gotcha in Go. The answer is really very clear in the docs:

> // The http Client and Transport guarantee that Body is always
    // non-nil, even on responses without a body or responses with
    // a zero-length body. It is the caller's responsibility to
    // close Body. The default HTTP client's Transport may not
    // reuse HTTP/1.x "keep-alive" TCP connections if the Body is
    // not read to completion and closed.

The part to stress on is _read to completion_. So therefore, just closing the Body is not sufficient to release the connection and the client may hold on to it if the body is not read.